File: mf-backend/src/modules/ledgers/service/ledger.service.ts
import { Injectable, NotFoundException, ForbiddenException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Ledger, LedgerStatus } from '../entity/ledger.entity.js';
import { CreateLedgerDto } from '../dto/create-ledger.dto.js';
import { TransactionType } from '../../transactions/entity/transaction.entity.js';

@Injectable()
export class LedgerService {
  constructor(
    @InjectRepository(Ledger)
    private readonly ledgerRepository: Repository<Ledger>,
  ) {}

  // 1. Criar uma nova conta de dívida
  async create(createLedgerDto: CreateLedgerDto, userId: string) {
    const ledger = this.ledgerRepository.create({
      ...createLedgerDto,
      ownerId: userId,
      // Se tem participante vinculado, fica pendente até o amigo aceitar
      status: createLedgerDto.participantId
        ? LedgerStatus.PENDING
        : LedgerStatus.ACCEPTED,
    });
    return await this.ledgerRepository.save(ledger);
  }

  // 2. Listar todas as contas onde o usuário está envolvido
  async findAll(userId: string) {
    return await this.ledgerRepository.find({
      where: [{ ownerId: userId }, { participantId: userId }],
      relations: ['transactions'], // Traz as transações para o cálculo
    });
  }

  // 3. Buscar uma conta específica e calcular o saldo total
  async findOneWithBalance(id: string, userId: string) {
    const ledger = await this.ledgerRepository.findOne({
      where: { id },
      relations: ['transactions'],
    });

    if (!ledger) throw new NotFoundException('Dívida não encontrada');

    // Validar se tem vinculo com a divida
    if (ledger.ownerId !== userId && ledger.participantId !== userId) {
      throw new NotFoundException(
        'Acesso negado, você não participa desta divida',
      );
    }

    // Lógica Matemática: Calcula o saldo a partir das transações
    const balance = ledger.transactions.reduce((acc, t) => {
      return t.type === TransactionType.DEBT
        ? acc + Number(t.amount)
        : acc - Number(t.amount);
    }, 0);

    return { ...ledger, balance };
  }

  // 4. Aceitar uma dívida (apenas o participante pode aceitar)
  async accept(id: string, userId: string) {
    const ledger = await this.ledgerRepository.findOne({ where: { id } });

    if (!ledger) throw new NotFoundException('Dívida não encontrada.');

    if (ledger.participantId !== userId) {
      throw new ForbiddenException('Apenas o participante pode aceitar esta dívida.');
    }

    if (ledger.status !== LedgerStatus.PENDING) {
      throw new ForbiddenException('Esta dívida não está pendente.');
    }

    ledger.status = LedgerStatus.ACCEPTED;
    return await this.ledgerRepository.save(ledger);
  }

  // 5. Recusar uma dívida (apenas o participante pode recusar)
  async reject(id: string, userId: string) {
    const ledger = await this.ledgerRepository.findOne({ where: { id } });

    if (!ledger) throw new NotFoundException('Dívida não encontrada.');

    if (ledger.participantId !== userId) {
      throw new ForbiddenException('Apenas o participante pode recusar esta dívida.');
    }

    if (ledger.status !== LedgerStatus.PENDING) {
      throw new ForbiddenException('Esta dívida não está pendente.');
    }

    ledger.status = LedgerStatus.REJECTED;
    return await this.ledgerRepository.save(ledger);
  }

  // 6. Deletar uma conta de dívida (apenas o owner)
  async delete(id: string, userId: string) {
    const ledger = await this.ledgerRepository.findOne({
      where: { id },
    });

    if (!ledger) {
      throw new NotFoundException('Dívida não encontrada.');
    }

    if (ledger.ownerId !== userId) {
      throw new NotFoundException(
        'Apenas o criador da dívida pode excluí-la.',
      );
    }

    await this.ledgerRepository.remove(ledger);
    return { message: 'Dívida excluída com sucesso.' };
  }
}

File: mf-backend/src/main/java/com/myfinance/ledger/LedgerService.java
package com.myfinance.ledger;

import com.myfinance.ledger.dto.CreateLedgerDTO;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import java.util.List;

@ApplicationScoped
public class LedgerService {

    @Transactional
    public Ledger create(CreateLedgerDTO dto, String userId) {
        Ledger ledger = new Ledger();
        ledger.title = dto.title;
        ledger.targetName = dto.targetName;
        ledger.ownerId = userId;
        ledger.participantId = dto.participantId;

        // Se tem participante, precisa de aceite
        if (dto.participantId != null && !dto.participantId.isEmpty()) {
            ledger.status = LedgerStatus.PENDING;
        } else {
            ledger.status = LedgerStatus.ACCEPTED;
        }

        ledger.persist();
        return ledger;
    }

    public List<Ledger> findAll(String userId) {
        return Ledger.findByOwnerOrParticipant(userId);
    }

    public Ledger findOne(String id, String userId) {
        Ledger ledger = Ledger.findById(id);
        if (ledger == null)
            return null;

        if (!ledger.ownerId.equals(userId) && !userId.equals(ledger.participantId)) {
            throw new SecurityException("Sem permissão");
        }
        return ledger;
    }

    @Transactional
    public Ledger accept(String id, String userId) {
        Ledger ledger = findOne(id, userId);
        if (ledger == null)
            return null;

        if (!userId.equals(ledger.participantId)) {
            throw new SecurityException("Apenas o participante pode aceitar a dívida.");
        }

        if (ledger.status != LedgerStatus.PENDING) {
            throw new IllegalStateException("Esta dívida não está pendente de aceitação.");
        }

        ledger.status = LedgerStatus.ACCEPTED;
        return ledger;
    }

    @Transactional
    public Ledger reject(String id, String userId) {
        Ledger ledger = findOne(id, userId);
        if (ledger == null)
            return null;

        if (!userId.equals(ledger.participantId)) {
            throw new SecurityException("Apenas o participante pode recusar a dívida.");
        }

        if (ledger.status != LedgerStatus.PENDING) {
            throw new IllegalStateException("Esta dívida não está pendente de aceitação.");
        }

        ledger.status = LedgerStatus.REJECTED;
        return ledger;
    }

    @Transactional
    public boolean delete(String id, String userId) {
        Ledger ledger = findOne(id, userId);
        if (ledger == null)
            return false;

        // Apenas o dono pode deletar (regra atual do NestJS)
        if (!ledger.ownerId.equals(userId)) {
            throw new SecurityException("Apenas o dono pode deletar.");
        }

        return Ledger.deleteById(id);
    }
}
